lib/improver/nbhood.py

```diff
@@ -431,21 +433,44 @@ def run(cube, radius):
             Cube containing the smoothed field after the square neighbourhood
             method has been applied.
         """
-        if isinstance(cube.data, np.ma.MaskedArray):
-            msg = ('Masked data is not currently supported in ',
-                   'SquareNeighbourhood.')
-            raise ValueError(msg)
+        masked_flag = False
         if np.any(np.isnan(cube.data)):
             msg = ('Data array contains NaNs which are not currently ',
                    'supported in SquareNeighbourhood.')
             raise ValueError(msg)
-        summed_up_cube = SquareNeighbourhood.cumulate_array(cube)
-        grid_cells_x, grid_cells_y = (
-            Utilities.get_neighbourhood_width_in_grid_cells(
-                summed_up_cube, radius, MAX_RADIUS_IN_GRID_CELLS))
-        neighbourhood_averaged_cube = (
-            SquareNeighbourhood.mean_over_neighbourhood(
-                summed_up_cube, grid_cells_x, grid_cells_y))
+        # If the data is masked, the mask will be processed as well as the
+        # original_data * mask array.
```

Comment:
We could slightly rejig this so that we have a consistent output e.g. always an iris CubeList, rather than somtimes a cubelist and sometimes a python list, and also do away with the else.

I've also renamed the mask_cube so I can extract it/reference it by name later.

```
cubes_to_sum = iris.cube.CubeList([cube])
if isinstance(cube.data, np.ma.MaskedArray):
   masked_flag = True
   mask_cube = cube.copy()
   mask_cube.rename('mask_data')
   mask_cube.data = np.logical_not(cube.data.mask.astype(int))
   cube.data = cube.data.data * mask_cube.data
   cubes_to_sum.append(mask_cube)
````